src/sidebar/oauth-auth.js

```diff
@@ -79,14 +80,20 @@ function auth($http, settings) {
 
   // Set a timeout to refresh the access token a few minutes before it expires.
   function refreshAccessTokenBeforeItExpires(tokenInfo) {
-    var delay = tokenInfo.expiresIn * 1000;
-
     // We actually have to refresh the access token _before_ it expires.
     // If the access token expires in one hour, this should refresh it in
     // about 55 mins.
-    delay = Math.floor(delay * 0.91);
+    var refreshAfter = tokenInfo.expiresAt.valueOf() * 0.91;
+
+    function refreshAccessTokenIfNearExpiry() {
+      if (Date.now() > refreshAfter) {
+        refreshAccessToken(tokenInfo.refreshToken);
+      } else {
+        refreshAccessTokenBeforeItExpires(tokenInfo);
```

Comment:
If it's not time to refresh the access token yet, check again in another 5 secs. I'm having `refreshAccessTokenBeforeItExpires()` call itself again like this, rather than just using `setInterval()`, because of the way it behaves when it _is_ time to send a refresh request (see [comment above](https://github.com/hypothesis/client/pull/384/files#r116516554))